File: 2018/day-7/solution.py
import networkx as nx
import re

graph = nx.DiGraph()
lines = open('input.txt', 'r').read().strip().splitlines()
steps = [re.findall(r'\s(\w{1})\s', line) for line in lines]
[graph.add_edge(step[0], step[1]) for step in steps]
sorted_steps = nx.lexicographical_topological_sort(graph)

print('Part 1:')
print(''.join(sorted_steps))

File: 2018/day-7/solution.rb
require 'rgl/adjacency'
require 'rgl/connected_components'
require 'rgl/dot'
require 'rgl/topsort'
require 'graphviz'

puts 'Part 1:'

input = File.open('input.txt').readlines
steps = input.map{ |s| s.scan(/\s(\w{1})\s/).flatten}
dg=RGL::DirectedAdjacencyGraph[]
steps.each{ |s1, s2| dg.add_edge s1, s2 }
available = []
s = ''
stb = dg.topsort_iterator.set_to_begin

min = stb.map{ |x| stb[x] }.min
available = stb.select{ |x| stb[x] == min }.keys
available = available.sort
next_available = available.shift
s += next_available

while true
  available += dg.adjacent_vertices(s[-1])
  min = available.map{ |x| stb[x] }.min
  available = available.select{ |x| stb[x] == min }
  available = available.sort
  next_available = available.shift
  s += next_available

  if s.length == steps.flatten.uniq.length
    break
  end
end

puts s


# nice little visualization of what's happening
dg.write_to_graphic_file('jpg')
